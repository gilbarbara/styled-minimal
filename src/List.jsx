import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import { ListStyles } from './utils/system';
import { sizeType } from './utils/types';

export const StyledList = styled.ul`
  ${ListStyles.base};

  > li {
    ${ListStyles.item}
  
    + li {
      ${ListStyles.itemSibling}
    }
  }
`;

const List = ({ children, ...props }) => (
  <StyledList {...props}>{children}</StyledList>
);

List.propTypes = {
  /** element type */
  as: PropTypes.oneOf(['ul', 'ol']),
  bordered: PropTypes.bool,
  children: PropTypes.node.isRequired,
  inline: PropTypes.bool,
  reversed: PropTypes.bool,
  size: sizeType,
  start: PropTypes.number,
  styleType: PropTypes.oneOfType([
    PropTypes.oneOf([
      'disc',
      'circle',
      'square',
      'decimal',
      'lower-alpha',
      'none',
    ]),
    PropTypes.string,
  ]),
  type: PropTypes.oneOf(['1', 'a', 'A', 'i', 'I']),
};

List.defaultProps = {
  as: 'ul',
  bordered: false,
  inline: false,
  size: 'md',
  styleType: 'none',
  type: '1',
};

export default List;
