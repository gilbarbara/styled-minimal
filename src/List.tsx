import * as React from 'react';
import styled, { css } from 'styled-components';

import { getTheme, px } from './utils/helpers';
import { baseStyles, sx, system } from './utils/system';

import { DLProps, OLProps, ULProps } from './types';

const stylesItem = (props: any) => {
  const { size = 'md' } = props;
  const { padding } = getTheme(props, 'list');

  return css`
    padding: ${px(padding[size])};
  `;
};

const stylesDD = (props: any) => {
  const { size = 'md' } = props;
  const { padding } = getTheme(props, 'list');

  return css`
    margin-left: ${px(padding[size])};
    padding: 0 ${px(padding[size])} ${px(padding[size])} 0;
  `;
};

const stylesSibling = (props: any) => {
  const { frame } = props;
  const { borderColor } = getTheme(props, 'list');

  if (!frame) {
    return undefined;
  }

  return css`
    border-top: 1px solid ${borderColor};
  `;
};

const stylesChildren = (props: any) => {
  const { as, frame, inline, styleType } = props;
  const { borderColor } = getTheme(props, 'list');
  const border = `1px solid ${borderColor}`;

  if (as === 'dl') {
    return css`
      dt {
        font-weight: 700;
        ${stylesItem}
        ${frame && inline && `border-bottom: ${border};`};
        ${frame && inline && `border-right: ${border};`};
        ${inline && `flex-basis: 30%;`};
      }

      dd {
        ${stylesDD};
        ${frame && inline && `border-bottom: ${border};`};
        ${inline && stylesItem};
        ${inline && `flex-basis: 70%;`};
        ${inline && `margin-left: 0;`};

        + dt {
          ${!inline && stylesSibling};
        }

        &:last-of-type {
          ${inline && `border-bottom: 0;`};
        }
      }

      dt:last-of-type {
        ${inline && `border-bottom: 0;`};
      }
    `;
  }

  return css`
    > li {
      ${frame && (as === 'ol' || styleType !== 'none' || inline) && `border-left: ${border};`};
      ${inline && 'flex: 1'};
      line-height: 1.2;
      ${stylesItem}

      + li {
        ${stylesSibling}
      }
    }
  `;
};

export const StyledList = styled('ul')<any>(
  props => {
    const { as, frame, inline, styleType } = props;
    const { borderColor, borderRadius } = getTheme(props, 'list');
    const border = `1px solid ${borderColor}`;

    return css`
      ${baseStyles};
      ${frame && `border: ${border}`};
      ${frame && `border-radius: ${px(borderRadius)}`};
      display: flex;
      flex-direction: ${inline ? 'row' : 'column'};
      flex-wrap: ${as === 'dl' && inline ? 'wrap' : 'nowrap'};
      margin: 0;
      ${as === 'ul' && styleType === 'none' ? 'padding: 0;' : ''};
      ${as === 'ul' ? `list-style-type: ${styleType};` : ''};

      ${stylesChildren};
    `;
  },
  sx,
  system,
);

export const DL = React.forwardRef<HTMLDListElement, DLProps>((props, ref) => (
  <StyledList as="dl" ref={ref} {...props} />
));

DL.defaultProps = {
  frame: true,
  inline: false,
  size: 'md',
};

export const OL = React.forwardRef<HTMLOListElement, OLProps>((props, ref) => (
  <StyledList as="ol" ref={ref} {...props} />
));

OL.defaultProps = {
  frame: false,
  inline: false,
  size: 'md',
};

export const UL = React.forwardRef<HTMLUListElement, ULProps>((props, ref) => (
  <StyledList as="ul" ref={ref} {...props} />
));

UL.defaultProps = {
  frame: true,
  inline: false,
  size: 'md',
  styleType: 'none',
};
