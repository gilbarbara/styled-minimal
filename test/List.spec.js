import React from 'react';
import { mount } from 'enzyme';

import List from '../src/List';

describe('List', () => {
  const wrapper = mount(
    <List>
      <li>Far far away, behind the word mountains, far from the countries Vokalia and Consonantia, there live the blind texts.</li>
      <li>Morbi leo risus</li>
      <li>Porta ac consectetur ac</li>
      <li>Vestibulum at eros</li>
    </List>
  );

  it('should render properly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
