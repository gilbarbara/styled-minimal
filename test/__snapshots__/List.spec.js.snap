// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List should render properly 1`] = `
.c0 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin: 0;
  padding: 0;
  list-style-type: none;
}

.c0 > li {
  padding: 0;
}

.c0 > li + li {
  border-top: none;
  margin-top: 12px;
}

<List
  as="ul"
  bordered={false}
  inline={false}
  size="md"
  styleType="none"
  type="1"
>
  <ul
    className="c0"
    size="md"
    type="1"
  >
    <li>
      Far far away, behind the word mountains, far from the countries Vokalia and Consonantia, there live the blind texts.
    </li>
    <li>
      Morbi leo risus
    </li>
    <li>
      Porta ac consectetur ac
    </li>
    <li>
      Vestibulum at eros
    </li>
  </ul>
</List>
`;
